DynamicTablesPkg/Library/Acpi/Common/AcpiSpcrLib/SpcrGenerator.c

```diff
@@ -107,6 +118,147 @@ GET_OBJECT_LIST (
+        (SerialTerminalInterruptInfo[Index].InterruptType !=
+         EFI_ACPI_SERIAL_PORT_CONSOLE_REDIRECTION_TABLE_INTERRUPT_TYPE_GIC))
+    {
+      DEBUG ((
+        DEBUG_ERROR,
+        "ERROR: SPCR: %d GIC interrupt type cannot be combined with others.\n",
+        SerialTerminalInterruptInfo[Index].InterruptType
+        ));
+      return EFI_INVALID_PARAMETER;
+    }
+
+    if ((SerialTerminalInterruptInfo[Index].InterruptType &
+         EFI_ACPI_SERIAL_PORT_CONSOLE_REDIRECTION_TABLE_INTERRUPT_TYPE_APIC) != 0)
+    {
+      if ((SerialTerminalInterruptInfo[Index].InterruptType &
+           ~(EFI_ACPI_SERIAL_PORT_CONSOLE_REDIRECTION_TABLE_INTERRUPT_TYPE_8259 |
+             EFI_ACPI_SERIAL_PORT_CONSOLE_REDIRECTION_TABLE_INTERRUPT_TYPE_APIC)) != 0)
+      {
+        DEBUG ((
+          DEBUG_ERROR,
+          "ERROR: SPCR: %d Invalid APIC interrupt type.\n",
+          SerialTerminalInterruptInfo[Index].InterruptType
+          ));
+        return EFI_INVALID_PARAMETER;
+      }
+    }
+
+    if ((SerialTerminalInterruptInfo[Index].InterruptType &
+         EFI_ACPI_SERIAL_PORT_CONSOLE_REDIRECTION_TABLE_INTERRUPT_TYPE_SAPIC) != 0)
+    {
+      if ((SerialTerminalInterruptInfo[Index].InterruptType &
+           ~(EFI_ACPI_SERIAL_PORT_CONSOLE_REDIRECTION_TABLE_INTERRUPT_TYPE_8259 | \
+             EFI_ACPI_SERIAL_PORT_CONSOLE_REDIRECTION_TABLE_INTERRUPT_TYPE_SAPIC)) != 0)
+      {
+        DEBUG ((
+          DEBUG_ERROR,
+          "ERROR: SPCR: %d Invalid SAPIC interrupt type.\n",
+          SerialTerminalInterruptInfo[Index].InterruptType
+          ));
+        return EFI_INVALID_PARAMETER;
+      }
+    }
+  }
+
+  return EFI_SUCCESS;
+}
+
+/** Update the SPCR table with the terminal interrupt type and terminal information.
+
+  @param [in, out] SpcrTable                   Pointer to the SPCR table.
+  @param [in]      SerialTerminalInterruptInfo Pointer to the Serial Terminal Interrupt Information.
+  @param [in]      SerialPortInfo              Pointer to the Serial Port Information.
+
+  @retval EFI_SUCCESS           The information was updated successfully.
+  @retval EFI_INVALID_PARAMETER A parameter is invalid.
+**/
+STATIC
+EFI_STATUS
+EFIAPI
+UpdateTerminalInterruptTypeInfo (
+  IN OUT EFI_ACPI_SERIAL_PORT_CONSOLE_REDIRECTION_TABLE_4  *SpcrTable,
+  IN     CM_ARCH_COMMON_SERIAL_TERMINAL_INTERRUPT_INFO     *SerialTerminalInterruptInfo,
+  IN     CM_ARCH_COMMON_SERIAL_PORT_INFO                   *SerialPortInfo
+  )
+{
+  if ((SpcrTable == NULL) || (SerialTerminalInterruptInfo == NULL)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  SpcrTable->InterruptType = SerialTerminalInterruptInfo->InterruptType;
+  SpcrTable->TerminalType  = SerialTerminalInterruptInfo->TerminalType;
+
+  if ((SerialTerminalInterruptInfo->InterruptType &
+       EFI_ACPI_SERIAL_PORT_CONSOLE_REDIRECTION_TABLE_INTERRUPT_TYPE_8259) != 0)
+  {
+    SpcrTable->Irq = SpcrTable->GlobalSystemInterrupt;

```

Comment:
I'm not really sure this configuration is actually possible, but is it possible to have `SpcrTable->Irq != SpcrTable->GlobalSystemInterrupt` ?